src/loader/loader.c

```diff
@@ -119,40 +111,91 @@ static shmem_request_hook_type prev_shmem_request_hook;
 #endif
 static ProcessUtility_hook_type prev_ProcessUtility_hook;
 
-/* This is timescaleDB's versioned-extension's post_parse_analyze_hook */
-static post_parse_analyze_hook_type extension_post_parse_analyze_hook = NULL;
+typedef struct TsExtension
+{
+	/*
+	 * Static data
+	 */
+
+	// Name of the extension (must be part of the so file name)
+	char const *const name;
+	// Name of the schema for table_name.
+	char const *const schema_name;
+	// Name of the table whose existence indicates the extension is loaded.
+	char const *const table_name;
+	// Name of the GUC for disabling loading this extension.
+	char const *const guc_disable_load_name;
+
+	/*
+	 * Run-time state
+	 */
+
+	// Current value of this extension's disable GUC.
+	bool guc_disable_load;
+
+	// Shared object library version loaded; empty if none.
+	char soversion[MAX_VERSION_LEN];
 
-inline static void extension_check(void);
+	// TODO Remove.  Neither timescaledb nor OSM actually have this hook,
+	// never have, and we don't plan to add them.
+	post_parse_analyze_hook_type post_parse_analyze_hook;
+} TsExtension;
+
+TsExtension extensions[] = {
+	// Redundant default initializers are here because we compile with
+	// `-Werror -Wmissing-field-initializers` for our PG13 build...
+	{
+		.name = "timescaledb",
+		.schema_name = CACHE_SCHEMA_NAME,
+		.table_name = EXTENSION_PROXY_TABLE,
+		.guc_disable_load_name = "timescaledb.disable_load",
+		.guc_disable_load = false,
+		.soversion = "",
+		.post_parse_analyze_hook = NULL,
+	},
+	{
+		.name = "timescaledb_osm",
+		.schema_name = "_osm_catalog",
+		.table_name = "metadata",
+		.guc_disable_load_name = "timescaledb_osm.disable_load",
+		.guc_disable_load = false,
+		.soversion = "",
+		.post_parse_analyze_hook = NULL,
+	},
+};
+
+inline static void extension_check(TsExtension *);
 #if PG14_LT
-static void call_extension_post_parse_analyze_hook(ParseState *pstate, Query *query);
+static void call_extension_post_parse_analyze_hook(ParseState *pstate, Query *query,
+												   TsExtension const *);
 #else
 static void call_extension_post_parse_analyze_hook(ParseState *pstate, Query *query,
-												   JumbleState *jstate);
+												   TsExtension const *, JumbleState *jstate);
 #endif
 
 static bool
-extension_is_loaded(void)
+extension_is_loaded(TsExtension const *const ext)
 {
 	/* The extension is loaded when the version is set to a non-null string */
-	return soversion[0] != '\0';
+	return ext->soversion[0] != '\0';
 }
 
 extern char *
 ts_loader_extension_version(void)
 {
-	return extension_version();
+	return extension_version(EXTENSION_NAME);
 }
 
 extern bool
 ts_loader_extension_exists(void)
 {
-	return extension_exists();
+	return extension_exists(EXTENSION_NAME);
 }
 
 static bool
```

Comment:
It would be good to add some commentary on what happens when an extension is dropped and how other backends connected to the same database pick up the change in the extension state.